### PR TITLE
[9351] Upload wheel + sdist on PyPI when a tag is created.

### PR DESCRIPTION
**Pull request by adiroiban**
# Scope

Based on the code from https://github.com/twisted/twisted/pull/1458

I have moved that code to Twisted repo to use the secrets

This is just to see that wheel + sdist are automatically published on PyPi when a new tag is crated.
Release documentation will be done at https://twistedmatrix.com/trac/ticket/10034

The release fails if the tag name doesn't match the branch version. Versions are PEP440.

The release is not waiting for tests to be green as the pypy3 run can take more than 22 minutes to run and randomly fail...and this will block the release.

The idea is that a tag is manually created by a person only after checking that all tests are green.

# Testing process

Step 1: Use PyPI testing and publish on every push to this PR
Step 2: Use PyPI testing and publish when a new tag is created - done for twisted-20.11.0.dev6

All steps were successfully executed. Result for last step at https://github.com/twisted/twisted/runs/1299338123?check_suite_focus=true

Final push - Use PROD PyPI and don't push any tag. We will test this on the first pre-release

## Contributor Checklist:

* [x] The associated ticket in Trac is here:  https://twistedmatrix.com/trac/ticket/9531
* [x] I ran `tox -e lint` to format my patch to meet the [Twisted Coding Standard](https://twistedmatrix.com/documents/current/core/development/policy/coding-standard.html)
* [ ] I ran additional checks listed at [Getting Your Patch Accepted](https://twistedmatrix.com/trac/wiki/TwistedDevelopment#GettingYourPatchAccepted)
* [x] I have created a newsfragment in src/twisted/newsfragments/ (see: [News files](https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles))
* [ ] I have updated the automated tests.
* [x] I have submitted the associated Trac ticket for review by adding the word `review` to the keywords field in Trac, and putting a link to this PR in the comment; it shows up in https://twisted.reviews/ now.

### CODE REVIEW

File: tox.ini
Comment (graingert) [inline tox.ini]:
```suggestion
    check-manifest>=0.44
```

File: tox.ini
Comment (graingert) [inline tox.ini:153]:
```suggestion
; Create sdist, extract wheel, and run basic tests on them.
```

**Comment by graingert** [comment]
lgtm other than some minor comments - although it's based on code I wrote so you might want to get an approval from someone else too

File: .github/workflows/test.yaml
Comment (graingert) [inline .github/workflows/test.yaml]:
```suggestion
      if: startsWith(github.ref, 'refs/tags')
      uses: pypa/gh-action-pypi-publish@master
      with:
        password: ${{ secrets.PYPI_UPLOAD_TOKEN }}
```

:shipit:

File: tox.ini
Comment (adiroiban) [reply tox.ini:153]: Well, even if the wheel is extracted from tox... the overall result of this env is that a wheel is created :)... how it is created, is low-level technical detail :)

File: .github/workflows/lint.yaml
Comment (graingert) [inline .github/workflows/lint.yaml:16]:
```suggestion
    steps:
    - uses: actions/checkout@v2
      with:
        # Need history on all branches for newsfragment and check-manifest to work
        fetch-depth: 0
```

**Review by graingert — CHANGES_REQUESTED** [review]
needs the fetch-depth hack on lint.yaml

File: .github/workflows/test.yaml
Comment (adiroiban) [reply .github/workflows/test.yaml]: I kept those comments to make it easy to switch to testing in a future branch.

File: .github/workflows/lint.yaml
Comment (adiroiban) [reply .github/workflows/lint.yaml:16]: True. My bad. Fixed.

File: admin/check_tag_version_match.py
Comment (graingert) [inline admin/check_tag_version_match.py]: you might want to use `pep517.meta.load(...).version`

Comment (adiroiban) [reply admin/check_tag_version_match.py]: done. thanks.

**Review by graingert — APPROVED** [review]
lgtm, I think this should get another reviewer though @exarkun ?

**Comment by adiroiban** [comment]
I am going to merge this as I hope it will reduce the load generated by wheel builders...and I think that this is on the right track.

**Comment by rodrigc** [comment]
@adiroiban This should not have passed review by @graingert, since this impacts the release process.  Unfortunately, I missed this review

File: tox.ini
Comment (rodrigc) [inline tox.ini:11]: This wheel target should not have been removed.

File: tox.ini
Comment (rodrigc) [inline tox.ini:153]: This should not have been removed

File: tox.ini
Comment (graingert) [reply tox.ini:11]: tox now builds the wheel before running any stage - it doesn't need a separate target to do it anymore

Comment (rodrigc) [reply tox.ini:11]: @graingert This affects the release process.  You and Adiroiban are not doing the current release (I am), so this is out of your scope to review, although technically the Twisted review process does not forbid this.

Comment (adiroiban) [reply tox.ini:11]: I think that we need the help from @glyph or @exarkun to weight their views on this.

I stand by my review and I think that @graingert did an excellent work
and this PR is a big improvement for the Twisted release process.

-----------

I know that the review process for this PR is on the grey area.

The code was created by @graingert in a fork.
To review and tests it, I have created this PR in a branch on the main repo.
So by creating this PR I was doing the review. 

----------

I might be wrong and I am not against reverting this PR if required.

Comment (adiroiban) [reply tox.ini:11]: Another argument for keeping this PR (inspired by Jean Paul, see irc logs https://freenode.logbot.info/twisted-dev/20210201#c6744337).

This argument represent only my views  and I am not representing JP or the Twisted community here :)
 
This PR is part of a long term goal of automating the release process.
The idea is that with this PR, you only need to create a tag, and the wheels are published by themselves :)
The release manager will not have to spend time on manually creating and publish the wheels.

The current release is in process since a very long time and there was not much progress.
I was feeling that by just following the old release process, we will continue to have delays.

So in parallel with the current release, some Twisted contributors (including myself) have tried to find a way to make the release work.

Cheers

Comment (rodrigc) [reply tox.ini:11]: You've done the following in the past few months:
• Made changes to the release automation
• Made changes to the release documentation
• Get people who are not doing the release, and have no track record doing Twisted releases to approve your commits

This behavior is very aggressive on your part.

While I've been a bit slow at this, on my part, I have gone out of my way to email Glyph and Amber, and
try to make sure that they were OK with me doing this release.  Unfortunately, I have not received any feedback from Amber, but Glyph gave me permission to do the release, so I have been proceeding with this.

I'm sorry that I have been a bit slow at releasing Twisted fast enough for you, but this is my first time doing this, and with the last issue fixed, I will do the release.

File: tox.ini
Comment (graingert) [reply tox.ini:153]: it was not removed, it was moved to `tox -e release-prepare`

File: tox.ini
Comment (adiroiban) [reply tox.ini:11]: I am not in a hurry to see a release or have it on PyPI.
For my personal project I use a fork and an private package index.

If you want to revert this, feel free to do it.

The changes that I made were not done in a hurry and I asked for feedback over IRC, beside the normal GitHub PR.

---------

I have worked with Thomas in the past on ldaptor and I know he is a good release manager and up to date with latest best practices in terms of Python package management and release process. 

I consider that he is very competent to do and review release tasks.

My suggestion is to move this conversation over the mailing list so that more people could weight their opinion. 

Thanks!

Comment (rodrigc) [reply tox.ini:11]: I would like to revert all these changes you have done to the release documentation and
process.

However, it looks like someone (probably you?)  has made a change to the GitHub Twisted repository, so
that a PR can only be merged to trunk, if it has been approved by someone who has write access
to the Twisted repository.   As someone with write access to the Twisted repository, in the past (as recent as summer 2020), it was possible
for me to get reviews approved by people who did not have write access to the repository.
This is not the case any more.
In private e-mail, Glyph said that he did not make this change to the GitHub repository, but he is in favor of keeping this configuration.  So I will respect that.

If I start reverting these changes, I will need to get review approval from another person with write access to the Twisted repository.  This will further delay the release, which I am not in favor of doing.

So in the interest of moving things along with the release, I will not be reverting these changes,
but I will express my disapproval.

I appreciate your enthusiasm for the Twisted project and all that you have done over the years, such as trying
to revive the Project Leadership Committee.
Unfortunately, in this area, I find that your recent behavior is very aggressive,
although it is entirely legal via the Twisted development process.

Comment (exarkun) [reply tox.ini:11]: Craig, can I suggest that you take the contributions of adiroiban and graingert to be in good faith?

Twisted is a group effort.  Right now, I see three people who want to help Twisted in this area.  I think that can be a great thing as long as we are working towards a common goal and not at cross purposes.

A lot of Twisted developers engage on IRC and on the mailing list which helps them come to a common goal and synchronize their efforts.

From your comments, I infer that you've had some private conversations with some people - but not with adiroiban or graingert.  Do you think you could start having some public conversations about what your goals and activities are?  I would suggest starting a thread on the Twisted mailing list rather than continuing to comment on this PR (which is public but obscure enough that a lot of people are going to miss out on the comments - as you yourself missed out on this PR when it was first submitted).  I think that until all interested parties are productively communicating with each other, it's going to be very hard to resolve this to everyone's satisfaction.

I hope you take this comment in the spirit it is given - as an attempt to constructively work through some obstacles that have arisen from lack of a shared goal and coordinated efforts towards that goal - not as an attack or any sort of aggression towards you.

Comment (rodrigc) [reply tox.ini:11]: @exarkun thanks for the feedback.  Yes, I believe that adiroiban and graingert have been doing things in good
faith, however, the way it has been done comes across as aggressive, and is unappreciated by me.

I think that I have expressed my views in this thread, in in other PR's, so I don't feel a need to
open another thread on the mailing list.

In terms of private conversations, I have reached out to Glyph and Amber (who did the last release).
I tried to be respectful of Amber's role, and make sure that I was not offending anyone by stepping up and trying to release Twisted.  Unfortunately, Amber has not responded to any of my emails, but Glyph has.

With respect to the release, for now, I am focusing on the last technical issue, and then cutting the release.

I think some of the root causes behind this problems we are seeing in this thread are due to the lack of an active Project Leadership Committee.  If there is anything you can do to help in that regard, especially in terms of answering these questions which I raised on the mailing list: https://twistedmatrix.com/pipermail/twisted-python/2021-January/065405.html , that would be great.

Comment (glyph) [reply tox.ini:11]: > I think some of the root causes behind this problems we are seeing in this thread are due to the lack of an active Project Leadership Committee. If there is anything you can do to help in that regard, especially in terms of answering these questions which I raised on the mailing list: https://twistedmatrix.com/pipermail/twisted-python/2021-January/065405.html , that would be great.

There's a lot going on in this discussion thread and I'll try to write a more substantive message later, but I just wanted to address this specific detail:

The Project "Leadership" Committee is perhaps unfortunately named.  This implies that the question you were asking about "roles" was not what I thought it was — about division of responsibilities within the group — but instead about what the group does as a whole.

So: the committee's sole responsibility is to *suggest that expenditures be approved* to the Software Freedom Conservancy, based on existing community consensus.  (I can't say "approve expenditures" because it doesn't have the unilateral ability to do so, its oversight is subject to the authority of the SFC itself.)  A better name for the group would be the Twisted Project Budgetary Oversight Committee.  Much as the Python Software Foundation does not manage the day-to-day operations of the Python core development team, even a highly efficient PLC would not have any role in addressing issues like this one, except in that it might hire someone to do the work.

So there's no authority coming to save us here, we need to build a community consensus through discussion of the issues on the merits.

**Pull request merged**
